clusterloader/docs/vision.md

```diff
@@ -0,0 +1,344 @@
+tests to that framework.
+
+We would like to get back to that and build fully functional Cluster Loader and
+use it as a framework to run all our scalability tests. This doc is describing
+high-level vision of how this will work.
+
+
+## Vision
+
+At the high level, a single test will consist of a number of steps. If we have
+really good usecase, we may consider creating a DAG of steps, but for now we
+don't have any rationale for it. In each of those steps we will be creating,
+updating and/or deleting a number of different objects.
+Additionally, we will introduce a set of predefined operations (that user will be
+able to use as phases). They will allow users to monitor/measure performance impact
+of user-defined phases.
+The following subsections will describe this in a bit more detail.
+
+### Config
+
+A single test scenario will be defined by a `Config`. Its schema will be as following:
+
+```
+struct Config {
+	// TBD: Some config allowing to access cluster, e.g. path to kubeconfig file.
+
+	// Number of namespace in the test.
+	Namespaces int32
+	// Steps of the test.
+	Steps []Step
+}
+```
+
+With a test being defined by a single json/yaml file, it should be pretty simple
+to modify scenarios, and fork them to new ones.
+
+Note that before running any steps, ClusterLoader will create all the necessary
+namespaces and after running all of them will delete them (together with all
+objects that remained undeleted after running the test). Namespaces are described
+in more details in the later part of this document.
+
+### Step
+
+Each step will consist of a number of create, update and delete operations (potentially
+on many different object types) or a number of monitoring/measurement-related actions.
+A single step is defined as following:
+
+```
+struct Step {
+	// Only one of these can be non-empty.
+	Phases       []Phase
+	Measurements []string
+}
+```
+
+We make `Phases` and `Measurements` separate concepts to ensure correct ordering
+between those two types of actions. It's very important to ensure that proper
+measurements are started before we start given actions and they are stopped when
+all actions are done.
+
+Also note that all `Phases` and `Measurements` within a single `Step` will be
+run in parallel - a `Step` ends when all its `Phases` or `Measurements` finish.
+
+### Phase
+
+Phase declaratively defines state of objects we should reach in the underlying
+Kubernetes cluster. A single declaration may result in a number of create, update
+and delete operations depending on the current state of the cluster.
+
+We define the phase as following:
+
+```
+struct Phase {
+	// Set of namespaces in which objects should be reconciled.
+	// If null, objects are assumed to be cluster scoped.
+	NamespaceRange *NamespaceRange
+	// Number of instances of a given object to exist in each
+	// of referenced namespaces.
+	ReplicasPerNamespace int32
+	// Name of TuningSet to be used.
+	TuningSet string
```

Comment:
You are using name of tuningSet instead of the struct itself. Where are we defining those tuningSets then? Are they defined globally (I can't find them in `Config`)?

Reply:
Good point - yes - they should be part of config. Added.